### PR TITLE
Kerberos Delegation feature for API Manger 3.0.0

### PR DESCRIPTION
**Pull request by AndreaNim**
In this PR,

- Add Kerberos delegation handler 

- Populate registry fields for Registry

- Add another type called "KERBEROS" for the API Endpoint Security
 
-  when creating an API ,In the swagger definition add  **endpointSecurity type "KERBEROS"** and traget SPN and **Authorization header as  X-APIM-Auth**
```
"endpointSecurity":{
     "type":"KERBEROS",
     "targetSpn":"service/example.com"

  },
```

- Create a directory called “kerberos” inside <APIM_HOME>/repository/conf/security/ folder and add the following files,

- krb5.conf
- login.conf
- keytab

Sample krb5.conf file

```
[libdefaults]
        default_realm = EXAMPLE.COM

[realms]
        EXAMPLE.COM = {
        kdc = kdc1.example.com
                  }

[domain_realm]
        .example.com = EXAMPLE.COM

```
Sample login.conf 
```
fileKrbLogin {
 com.sun.security.auth.module.Krb5LoginModule required
 useTicketCache=false
 refreshKrb5Config=true
 doNotPrompt=true
 useKeyTab=true
 debug=true
 storeKey=true
 principal="< principal >"
 keyTab="< keytab_path >";
};
```
Command to create a Keytab
`ktpass /out <keytab> /princ HTTP/<FQDN>@<DOMAIN NAME>/mapuser  <User Name of the AD account> /pass <password> /crypto All /ptype KRB5_NT_PRINCIPAL
`

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/API.java [hunk 1]
```diff
@@ -90,6 +90,9 @@
     private String endpointUTUsername;
     private String endpointUTPassword;
 
+    private boolean endpointKerberosEnable= false;
```
Comment (hasuniea) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/API.java]: fix the formatting issue.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+            log.debug("Initiating constrained delegation for SPN: " + targetSpn);
+        }
+        byte[] delegatedTicket = kerberosDelegator
+                .delegate(Base64.getDecoder().decode(clientKerberosTicket.split(" ")[1]), targetSpn);
+        String delegatedKerberosTicket = Base64.getEncoder().encodeToString(delegatedTicket);
+        if (log.isDebugEnabled()) {
+            log.debug("Acquired delegated Kerberos ticket: " + delegatedKerberosTicket);
+        }
+        return delegatedKerberosTicket;
+
```
Comment (hasuniea) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java]: remove the new line.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/mappings/APIMappingUtil.java [hunk 3]
```diff
@@ -487,6 +487,11 @@ private static void setEndpointSecurityFromApiDTOToModel(APIDTO dto, API api) {
         APIEndpointSecurityDTO securityDTO = dto.getEndpointSecurity();
         if (dto.getEndpointSecurity() != null && securityDTO.getType() != null) {
             api.setEndpointSecured(true);
+            if(APIEndpointSecurityDTO.TypeEnum.KERBEROS.equals(securityDTO.getType())){
```
Comment (hasuniea) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/mappings/APIMappingUtil.java]: format the if condition and remove the new line.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/UserManager.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
+/*
+ *
+ *  * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (hasuniea) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/UserManager.java]: Please correct the license header

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosConstants.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    public static final String AUTHENTICATE_HEADER = "WWW-Authenticate";
+    public static final String LOGIN_CONFIG_PROPERTY = "java.security.auth.login.config";
+    public static final String LOGIN_CONF_FILE_NAME = "login.conf";
+    public static final String KERBEROS_CONFIG_PROPERTY = "java.security.krb5.conf";
+    public static final String KERBEROS_CONF_FILE_NAME = "krb5.conf";
+    public static final String LOGIN_CONTEXT_NAME = "loginContextName";
+    // Negotiate header string.
+    public static final String NEGOTIATE = "Negotiate";
+
+
```
Comment (hasuniea) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosConstants.java]: remove new lines.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+import org.wso2.carbon.apimgt.gateway.handlers.security.kerberos.utils.KerberosDelegator;
+import org.wso2.carbon.apimgt.gateway.handlers.security.kerberos.utils.UserManager;
+import org.wso2.carbon.utils.CarbonUtils;
+
+import java.nio.file.Paths;
+import java.util.Base64;
+import java.util.Map;
+import javax.security.auth.login.Configuration;
+
+public class KerberosDelegationHandler extends AbstractHandler {
```
Comment (hasuniea) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java]: Add java doc comment.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/KerberosDelegator.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    private GSSContext startAsClient(final String targetSPN, final GSSCredential delegationCredential) throws Exception {
+        PrivilegedExceptionAction<GSSContext> action = new PrivilegedExceptionAction<GSSContext>() {
+            @Override
+            public GSSContext run() throws Exception {
+                GSSManager manager = GSSManager.getInstance();
+                GSSName serverName = manager.createName(targetSPN, krb5PrincipalNameType);
+                GSSContext gssContext = manager
+                        .createContext(serverName.canonicalize(spnegoOid), spnegoOid, delegationCredential,
+                                GSSContext.DEFAULT_LIFETIME);
+                //gssContext.requestMutualAuth(true);
```
Comment (ChamodDamitha) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/KerberosDelegator.java]: Shall we remove this commented lines

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 8]
```diff
@@ -1977,7 +1977,17 @@ private boolean isAPIPublished(API api) throws APIManagementException {
     }
 
     private APITemplateBuilder getAPITemplateBuilder(API api) throws APIManagementException {
+
         APITemplateBuilderImpl vtb = new APITemplateBuilderImpl(api);
+
```
Comment (ChamodDamitha) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Remove newline

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/dto/APIEndpointSecurityDTO.java [hunk 9]
```diff
@@ -69,6 +70,23 @@ public void setType(TypeEnum type) {
     this.type = type;
   }
 
+  /**
+   **/
+  public APIEndpointSecurityDTO targetSpn(String targetSpn) {
+    this.targetSpn = targetSpn;
+    return this;
+  }
+
```
Comment (ChamodDamitha) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/dto/APIEndpointSecurityDTO.java]: remove newline

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/mappings/APIMappingUtil.java [hunk 10]
```diff
@@ -487,6 +487,11 @@ private static void setEndpointSecurityFromApiDTOToModel(APIDTO dto, API api) {
...
+                api.setEndpointKerberosEnable(true);
+                api.setEndpointTargetSpn(securityDTO.getTargetSpn());
+
```
Comment (ChamodDamitha) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/mappings/APIMappingUtil.java]: Remove newline

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+
+    private boolean isUserAuthorized(Map headersMap, MessageContext messageContext, byte[] serverToken) {
+        org.apache.axis2.context.MessageContext axis2MessageContext = ((Axis2MessageContext) messageContext)
+                .getAxis2MessageContext();
+        String outServerTokenString = null;
+        headersMap.clear();
+            if (serverToken != null) {
+                outServerTokenString = Base64.getEncoder().encodeToString(serverToken);
+            }
+            axis2MessageContext.setProperty("HTTP_SC", "401");
```
Comment (HiranyaKavishani) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java]: Shall we assign this property values as constant variables?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/UserManager.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+     * Login this user using the given login context.
+     *
+     * @param loginContextName login context name
+     * @throws Exception if login fails
+     */
+    public void login(String loginContextName) throws Exception {
+        if (isLoggedin()) {
+            throw new IllegalStateException("User '" + getSubject().toString() + "' is already logged in.");
+        }
+
```
Comment (HiranyaKavishani) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/UserManager.java]: Remove new line

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/API.java (hunk 1 above)
Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/API.java]: fixed.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java (hunk 2 above)
Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java]: removed the new line.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/mappings/APIMappingUtil.java (hunk 3 above)
Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/mappings/APIMappingUtil.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/UserManager.java (hunk 4 above)
Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/UserManager.java]: Added the correct licence header.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosConstants.java (hunk 5 above)
Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosConstants.java]: removed the new line.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/mappings/APIMappingUtil.java (hunk 3 above)
Comment (HiranyaKavishani) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/mappings/APIMappingUtil.java]: Swap the condition as `(securityDTO.getType(). equals (APIEndpointSecurityDTO.TypeEnum.KERBEROS))`

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java (hunk 6 above)
Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java]: added the comment.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/KerberosDelegator.java (hunk 7 above)
Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/KerberosDelegator.java]: removed them.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 8 above)
Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: removed the newline.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/dto/APIEndpointSecurityDTO.java (hunk 9 above)
Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/gen/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/dto/APIEndpointSecurityDTO.java]: removed the newline.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/mappings/APIMappingUtil.java (hunk 10 above)
Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/mappings/APIMappingUtil.java]: removed the newline

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+            log.info("krb5 debug logs enabled.");
+            System.setProperty("sun.security.krb5.debug", "true");
+            System.setProperty("sun.security.jgss.debug", "true");
+        }
+
+        Map headers = getTransportHeaders(messageContext);
+        if (getKerberosHeader(headers) == null) {
+            return isUserAuthorized(headers, messageContext, null);
+        } else {
+
```
Comment (HiranyaKavishani) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java]: Remove new line

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+                kerberosTicket = getDelegatedTicket(self, headers);
+                setKerberosHeader(headers, kerberosTicket);
+            } catch (Exception e) {
+                throw new RuntimeException("Kerberos constrained delegation failed with login context name." + loginContextName, e);
+            } finally {
+                self.logout();
+                // Revert back to previous configs
+                Configuration.setConfiguration(null);
+                if (jaasConfigPathold != null) {
+                    System.setProperty(KerberosConstants.LOGIN_CONFIG_PROPERTY, jaasConfigPathold);
```
Comment (HiranyaKavishani) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java]: remove new line

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/KerberosDelegator.java [hunk 15]
```diff
@@ -62,15 +62,14 @@ public KerberosDelegator(Subject selfSubject) {
...
             }
             if (delegationCredential != null) {
                 try {
                     delegationCredential.dispose();
                 } catch (GSSException e) {
-                    log.error("ERROR",e);
+                    log.error("Error while disposing GSS credential", e);
                     throw e;
                 }
             }
```
Comment (Shehanir) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/KerberosDelegator.java]: remove new lines

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/KerberosDelegator.java [hunk 16]
```diff
@@ -99,10 +98,9 @@ public GSSCredential run() throws Exception {
                         try {
                             context.dispose();
                         } catch (GSSException e) {
-                            log.error("ERROR",e);
+                            log.error("Error,failed to dispose GSS credential ", e);
                             throw e;
```
Comment (Shehanir) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/KerberosDelegator.java]: remove new lines

Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/KerberosDelegator.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java (hunk 14 above)
Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java (hunk 13 above)
Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/KerberosDelegationHandler.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/UserManager.java (hunk 12 above)
Comment (AndreaNim) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/kerberos/utils/UserManager.java]: fixed

**Pull request merged**